lib/usd/translators/shading/mtlxMaterialXSurfaceShaderWriter.cpp

```diff
@@ -0,0 +1,516 @@
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+#include "shadingTokens.h"
+
+#include <mayaUsd/fileio/shaderWriter.h>
+#include <mayaUsd/fileio/shaderWriterRegistry.h>
+
+#include <pxr/base/tf/token.h>
+#include <pxr/pxr.h>
+#include <pxr/usd/usdMtlx/utils.h>
+#include <pxr/usd/usdShade/material.h>
+#include <pxr/usd/usdShade/nodeGraph.h>
+#include <pxr/usd/usdUI/nodeGraphNodeAPI.h>
+
+#include <maya/MFnDependencyNode.h>
+#include <maya/MPlug.h>
+#include <ufe/path.h>
+#include <ufe/pathString.h>
+#include <ufe/runTimeMgr.h>
+
+#include <MaterialXFormat/XmlIo.h>
+
+PXR_NAMESPACE_OPEN_SCOPE
+
+// Retrieves the standard library document for MaterialX.
+MaterialX::ConstDocumentPtr _GetStandardLibraryDocument()
+{
+    static auto standardDoc = UsdMtlxGetDocument("");
+    return standardDoc;
+}
+
+// Gets the node definition string for a given node and UFE path.
+// If the Node already has it's nodeDefString set, use that.
+// Otherwise, get the nodeDef via UFE.
+std::string _GetNodeDefString(const MaterialX::NodePtr& node, const Ufe::Path& ufePath)
+{
+    auto nodeDefString = node->getNodeDefString();
+    if (!nodeDefString.empty()) {
+        return nodeDefString;
+    }
+
+    auto nodeDefHandler = Ufe::RunTimeMgr::instance().nodeDefHandler(ufePath.runTimeId());
+    auto sceneItem = Ufe::Hierarchy::createItem(ufePath);
+    auto nodeDef = nodeDefHandler->definition(sceneItem);
+    if (nodeDef) {
+        return nodeDef->type();
+    }
+    TF_WARN("Could not find nodeDef for node '%s'", node->getName().c_str());
+    return std::string();
+}
+
+// Retrieves the node definition for a given node and UFE path.
+MaterialX::ConstNodeDefPtr _GetNodeDef(const MaterialX::NodePtr& node, const Ufe::Path& ufePath)
+{
+    auto nodeDefName = _GetNodeDefString(node, ufePath);
+    auto nodeDefPtr = node->getDocument()->getNodeDef(nodeDefName);
+    if (!nodeDefPtr) {
+        // Get the standard library document and check that.
+        nodeDefPtr = _GetStandardLibraryDocument()->getNodeDef(nodeDefName);
+    }
+    return nodeDefPtr;
+}
+
+// Sets shader info:id attribute on a USD prim based on a MaterialX node.
+void _SetShaderInfoAttributes(
+    const MaterialX::NodePtr& node,
+    UsdPrim&                  shaderPrim,
+    const Ufe::Path&          ufePath)
+{
+    auto infoAttr = shaderPrim.CreateAttribute(
+        UsdShadeTokens->infoId, SdfValueTypeNames->Token, SdfVariabilityUniform);
+
+    auto nodeDefString = _GetNodeDefString(node, ufePath);
+    infoAttr.Set(TfToken(nodeDefString));
```

Comment:
That works fine, but it would be preferable to use the UsdShadeShader API:
``` 
    auto shader = UsdShadeShader(shaderPrim);
    shader.CreateIdAttr(VtValue(TfToken(nodeDefString)));
```